metadata-ingestion/src/datahub/ingestion/source/powerbi/proxy.py

```diff
@@ -0,0 +1,912 @@
+            LOGGER.warning(f"{Constant.ReportId}={report_id}")
+            raise ConnectionError(message)
+
+        response_dict = response.json()
+
+        return PowerBiAPI.Report(
+            id=response_dict.get("id"),
+            name=response_dict.get("name"),
+            webUrl=response_dict.get("webUrl"),
+            embedUrl=response_dict.get("embedUrl"),
+            description=response_dict.get("description"),
+            users=[],
+            pages=[],
+            dataset=self.get_dataset(
+                workspace_id=workspace_id, dataset_id=response_dict.get("datasetId")
+            ),
+        )
+
+    def get_access_token(self):
+        if self.__access_token != "":
+            LOGGER.info("Returning the cached access token")
+            return self.__access_token
+
+        LOGGER.info("Generating PowerBi access token")
+
+        auth_response = self.__msal_client.acquire_token_for_client(
+            scopes=[PowerBiAPI.SCOPE]
+        )
+
+        if not auth_response.get("access_token"):
+            LOGGER.warning(
+                "Failed to generate the PowerBi access token. Please check input configuration"
+            )
+            raise ConfigurationError(
+                "Powerbi authorization failed . Please check your input configuration."
+            )
+
+        LOGGER.info("Generated PowerBi access token")
+
+        self.__access_token = "Bearer {}".format(auth_response.get("access_token"))
+
+        LOGGER.debug(f"{Constant.PBIAccessToken}={self.__access_token}")
+
+        return self.__access_token
+
+    def get_dashboard_users(self, dashboard: Dashboard) -> List[User]:
+        """
+        Return list of dashboard users
+        """
+        return self.__get_users(
+            workspace_id=dashboard.workspace_id, entity="dashboards", _id=dashboard.id
+        )
+
+    def get_dashboards(self, workspace: Workspace) -> List[Dashboard]:
+        """
+        Get the list of dashboard from PowerBi for the given workspace identifier
+
+        TODO: Pagination. As per REST API doc (https://docs.microsoft.com/en-us/rest/api/power-bi/dashboards/get
+        -dashboards), there is no information available on pagination
+        """
+        dashboard_list_endpoint: str = PowerBiAPI.API_ENDPOINTS[Constant.DASHBOARD_LIST]
+        # Replace place holders
+        dashboard_list_endpoint = dashboard_list_endpoint.format(
+            POWERBI_BASE_URL=PowerBiAPI.BASE_URL, WORKSPACE_ID=workspace.id
+        )
+        # Hit PowerBi
+        LOGGER.info(f"Request to URL={dashboard_list_endpoint}")
+        response = requests.get(
+            dashboard_list_endpoint,
+            headers={Constant.Authorization: self.get_access_token()},
+        )
+
+        # Check if we got response from PowerBi
+        if response.status_code != 200:
+            LOGGER.warning("Failed to fetch dashboard list from power-bi for")
+            LOGGER.warning(f"{Constant.WorkspaceId}={workspace.id}")
```

Comment:
assuming much of this logic is copied from the original powerbi file

Reply:
Yes complete code, for current work we haven't added any API.